### ISSUE TITLE
14.0.1 breaks ssh-agent

### ISSUE THREAD

**Issue by tjim** [blinksh/blink#1291]
### Checklist

- [x] Using latest version on the [App Store](http://itunes.apple.com/app/id1156707581)
- [x] Read the [docs](http://blink.sh/docs/)
- [x] Searched for [existing GitHub issues](https://github.com/blinksh/blink/issues)

### Configuration

Blink version 14.0.1, iOS version 14.8, iPad

### Describe the bug

After updating ssh-agent does not work:

    blink> ssh-agent
    Failed loading ssh_agent_main from MAIN, cause = (null)
    ssh-agent: command not found

**Comment by carloscabanero** [comment]
The agent in 14 is now special and there is no need to run it separately. Just do a ssh-add and you should be able to get going ;)

**Comment by tjim** [comment]
OK but then you have broken ssh-add, which only seems to be able to add the key id_rsa now. I have several keys and I used to be able to add them by passing an argument to ssh-add, e.g.,

    ssh-add ~/.ssh/otherkey 

This now fails with

    Key not found

**Comment by tjim** [comment]
Also, the error message for ssh-agent is not great:

    Failed loading ssh_agent_main from MAIN, cause = (null)

**Comment by carloscabanero** [comment]
We are moving away from keys in the file system. I will reopen so we can change the docs on this.

Please import the key inside settings (config > Keys > import). Then the agent will let you do the rest.

**Comment by kokofirstlove** [comment]
It works!

ΣℓĢÂ฿ẹℓᵧ

في الخميس، ١٤ أكتوبر ٢٠٢١ ٥:٤٥ م carloscabanero ***@***.***>
كتب:

**Comment by tjim** [comment]
A problem with adding the keys via (config->keys->import) is that the usual file location for keys is in the .ssh directory and the “import from file” dialog does not see hidden folders (starting with .ssh).

If I move keys around I can import, but then authentication does not work for me.  Using ssh -vvv to try to debug is not very useful because it does not tell you what key is being offered.  For example in the debug output you see

    ssh_userauth_agent: Trying identity 
    …
    ssh_userauth_agent: Public key of  refused by server

Now there is space after “identity” above and two spaces in “of  refused” where I think it would normally print out the key that is being offered, but obviously that is failing.

**Comment by tjim** [comment]
Also: in the old behavior, ssh -A worked (forwarding the authentication agent), is that still going to work with the new ssh-agent functionality?

**Comment by carloscabanero** [comment]
Yes, that still works. The new agent is a lot more flexible and will help us, for example, to forward specific keys only to specific hosts, or add constraints to ask you before using a key. And solves all configuration and startup pains.

**Comment by rotecodefraktion** [comment]
After update, all Keys are gone. So does it mean we‘ve to load all again?

**Comment by carloscabanero** [comment]
@rotecodefraktion we performed a migration of files in Blink to a new container. There may be a bug on iOS15 where the app starts in the background, and the app has no access to files or anything, so the migration will fail. Keys and everything are not lost, but you may have to just complete it manually. Please check this https://github.com/blinksh/blink/issues/1294#issuecomment-944172177. We are working on a fix as we speak, very sorry for this :(

**Comment by carloscabanero** [comment]
@tjim Thanks, I will fix the issue of the key not showing properly. Did you get everything else up and running?

**Comment by rotecodefraktion** [comment]
> We are working on a fix as we speak, very sorry for this :(
@carloscabanero no worry about my loss. It was just a suprise after AutoUpdate to 15.0.2

**Comment by tjim** [comment]
> @tjim Thanks, I will fix the issue of the key not showing properly. Did you get everything else up and running?

No keys were working on upgrade, including authentication with the default id_rsa. After I deleted id_rsa from the config and re-added it, authentication with id_rsa works again. However, I’m not able to get authentication with other keys to work, and due to the logging issue I describe above I haven’t been able to debug — I’m not sure what keys are being offered during login. One thing that might be different for me than the others on the thread — I am still on iOS14.

**Comment by carloscabanero** [comment]
@tjim Im fixing the issue with the keys not showing up on verbose. To be sure or enforce a key, just set it up with "-i" flag. Please note that keys must be imported inside Config > Keys.

**Comment by tjim** [comment]
I’m now on Blink 14.0.2.354, I observe the following.

With multiple keys (id_rsa and some others) I have to use -i keyname if I need a non-default (non id_rsa) key. Previously (and consistent with openssh), ssh would try the different keys without -i, so this seems a downgrade.

Also, previously the -i flag took a file name as argument, e.g., -i ~/.ssh/keyfile. This no longer works. Instead, you must use just “keyfile”, that is, the name of the file when you imported it to blink, and what shows up as the name of the key in blink. This is inconsistent with the openssh documentation for -i, so this behavior ought to be documented somewhere.

Finally, the -A flag (agent forwarding) has different behavior. I need to log in to host A with key KA, and from there I log in to host B with key KB. I am able to ssh -i KA A, however, when I try to ssh B, I get an error, “Agent admitted failure to sign using the key.” If I try ssh -i KB B, I get an additional message, “Warning: Identity file B not accessible: No such file or directory.” This makes sense because the ssh on host A is not the blink ssh, it is openssh.

**Comment by carloscabanero** [comment]
Few things, just to clarify the behavior:

> With multiple keys (id_rsa and some others) I have to use -i keyname if I need a non-default (non id_rsa) key. Previously (and consistent with openssh), ssh would try the different keys without -i, so this seems a downgrade.

I think this is not correct, and it is actually a potential security issue. Previously and OpenSSH **only** tries the default key names (id_XXXX), and not "different keys". If we tried all the keys you have, you would not be able to "know" which one gets used, and it could be a problem as keys can be associated to specific commands. 

On the behavior of -i reading a key from a file and that being inconsistent with OpenSSH. We are phasing out file based keys. It is true that our implementation has worked with Blink installed keys as well as file based for a long time. Unfortunately this caused a ton of issues, debugging headaches and unexpected behaviors. We are reverting back to just "one way of doing things". For our future stuff this is what makes most sense as well.

On the different behavior, not sure I follow exactly what you mean. This should work though:
```
ssh-add KB
ssh -A -i KA A
ssh-add -l  # this should list two keys, KA which is inaccessible and KB, which is forwarded
ssh B # you do not need the KB here, because it is using a forwarded agent key, as you mention you do not have that file
```

**Comment by tjim** [comment]
> Few things, just to clarify the behavior:
> 
> > With multiple keys (id_rsa and some others) I have to use -i keyname if I need a non-default (non id_rsa) key. Previously (and consistent with openssh), ssh would try the different keys without -i, so this seems a downgrade.
> 
> I think this is not correct, and it is actually a potential security issue.

So it is a little confusing and turns out it depends on the state of ssh-agent and the interplay with ssh-add.

I had thought that within Blink, it was no longer necessary to run ssh-agent (as that is now built in), and, furthermore, since I added keys to blink via the settings, I thought that I no longer needed to do ssh-add. That is, I thought that adding a key via the Blink settings was equivalent to ssh-add.

And it turns out that is sort of true, and sort of not true.

Suppose we have host HA accepting key KA and host HB accepting key KB and both KA (== id_rsa) and KB are loaded into Blink via settings.

Then in Blink, ssh -A HA succeeds and on HA, ssh-add -l shows that key KA is available via the agent.

However in openssh with no keys loaded into ssh-agent, ssh -A HA also succeeds but on HA, ssh-add -l shows *NO* keys available via the agent.

In Blink, ssh -A HB fails. And ssh -A -i KB HB succeeds and on HB shows that key KB is available via the agent.

In openssh, ssh -A HB fails, and ssh -A -i KB HB succeeds and shows on HB that *NO* keys are available via the agent.  

Now this has all been happening without ssh-add.

In Blink if you ssh-add KA and ssh-add KB, then ssh -A HA and ssh -A HB *both* succeed. Note that the -i flag was not needed for HB, because the agent is able to pick out the correct key. This is exactly the behavior that I was referring to when I said “ssh would try the different keys without -i” above.  (And, I don’t think it is a security problem.) Also, on HA and HB, ssh-add -l shows that both KA and KB are available via the agent.

In openssh if you ssh-add KA and ssh-add KB, then both ssh -A HA and ssh-A HB work as well, and on HA and HB, both keys are available via the agent.

The summary is that the ssh-agent of Blink operates differently than that of openssh when *no* keys have been added with ssh-add, and ssh-add must be used in Blink even though keys are “added” via settings.  When this is done for multiple keys you get behavior much like openssh (where multiple keys are offered without requiring the -i flag).

Now that I understand the current behavior I can live with it.

The only question is whether this is going to confuse others — and I think it will. Maybe you should document it somewhere, or, you should try to make the behavior closer to openssh so that you can refer to their documentation?

Thanks!

**Comment by carloscabanero** [comment]
> I had thought that within Blink, it was no longer necessary to run ssh-agent (as that is now built in), and, furthermore, since I added keys to blink via the settings, I thought that I no longer needed to do ssh-add. That is, I thought that adding a key via the Blink settings was equivalent to ssh-add.

You no longer have to run the agent, but we did not say that you no longer had to add the keys **that you want to forward,** that would be a security issue and not standard at all. The OpenSSH agent does not forward keys on its own.

And that is what you are seeing and what I was explaining. After connecting, if you do `ssh-add -l`, you will see your key there, but it won't be used for signing because **it is constrained.** If you want to use it, for now you have to explicitly add it. We could hide it, like I think the Mac agent does, but we have decided not to yet.

Another reason why we did that, which I think fits with the model you want to get to. In the future, this offers a very simple path to implement the "AddKeysToAgent" function from our side. The IdentityFile at config allows to select "multiple keys", and we could potentially remove the constraints AFTER asking the user, something like:
- Host X wants to use key Y. Do you want to allow it?
* Just once
* Always for this session
* Deny

So to sum up, we need to rewrite parts of the help as they have changed, but I do not think this is out of the standard really. The confusing part may have been seeing a key in the remote that you are not able to use - it would be more correct to hide - but it is just not usable for signing. We want it there though, because we want you to be able to select "multiple keys" at one point, and then we can ask you the first time you use. I think I like how this agent will behave a lot more.

**Comment by carloscabanero** [comment]
Docs updated.
